### DISCUSSION TITLE
Detect nodejs vs edge runtime

### DISCUSSION THREAD

**Discussion by controversial** [vercel/next.js#53869] [Help]
### Summary

Is there a standard / recommended way to detect whether some chunk of code is running in the `nodejs` runtime or the `edge` runtime? An ideal solution would work within dev server too. 

This would be useful, for example, if there’s two implementations of a function that I want to choose between based off of the edge runtime tradeoffs (e.g. optimize for lower initial instantiation time on edge runtime)

**Comment by v1s10n-4** [comment]
If you're using vercel for deployments you can check the global variable `EdgeRuntime`, as stated in [their docs](https://vercel.com/docs/functions/edge-functions/edge-runtime#check-if-you're-running-on-the-edge-runtime).

``` js
if (typeof EdgeRuntime !== 'string') {
  // dead-code elimination is enabled for the code inside this block
}
```

**Comment by JamieS1211** [comment]
Any info on how to test this outside of a Vercel deployment? Sadly I cannot use Vercel due to compliance issues but still need to test if things are running on the edge or not.

**Comment by rubby-c** [comment]
I'd also like to know.

**Comment by BrettHoutz** [comment]
You can use `process.env.NEXT_RUNTIME`, as mentioned in the docs here:
https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation#importing-runtime-specific-code

(Not a great location for this documentation IMO. Those who want to use this for middleware would not necessarily look at the instrumentation page.)

**Comment by BearToCode** [comment]
The following does not work on SvelteKit. Has anyone found a solution?

```ts
// hooks.server.ts

export const handle: Handle = async ({ event, resolve }) => {
  if (typeof EdgeRuntime !== 'string') {
     await import('some-node-module');
  }
  return resolve(event);
}
```

`sone-node-module` is still imported!

**Reply by BearToCode** [reply]
# IMPORTANT

Dead code is not removed, as stated in the docs. If you need to import something `hooks.server.ts`, even thought it is imported dynamically and only if `typeof EdgeRuntime !== 'string'`, il will still give an error. You need to add all the builtin packages into the `external` setting of the `@sveltejs/adapter-vercel`!

**Comment by leerob** [comment]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
